pretixprint/app/src/main/java/eu/pretix/pretixprint/byteprotocols/BrotherRaster.kt

```diff
@@ -0,0 +1,193 @@
+class BrotherRaster : StreamByteProtocol<Bitmap> {
+    override val identifier = "BrotherRaster"
+    override val nameResource = R.string.protocol_brother
+    override val defaultDPI = 300
+    override val demopage = "demopage_8in_3.25in.pdf"
+
+    // Label data list extracted from
+    // https://download.brother.com/welcome/docp100366/cv_ql1100_eng_raster_100.pdf
+    enum class Label(val id: Int, val width: Int, val height: Int, val continuous: Boolean, val printableWidth: Int, val printableHeight: Int, val twoColor: Boolean = false) {
+        c12mm(257, 12, 0, true, 106, 0),
+        c29mm(258, 29, 0, true, 306, 0),
+        c38mm(264, 38, 0, true, 413, 0),
+        c50mm(262, 50, 0, true, 554, 0),
+        c54mm(261, 54, 0, true, 590, 0),
+        c62mm(259, 62, 0, true, 696, 0),
+        c62mm_rb(259, 62, 0, true, 696, 0, true),
+        c102mm(260, 102, 0, true, 1164, 0),
+        c103mm(265, 103, 0, true, 1200, 0),
+        d17x54(269, 17, 54, false, 165, 566),
+        d17x87(270, 17, 87, false, 165, 956),
+        d23x23(370, 23, 23, false, 236, 202),
+        d29x42(358, 29, 42, false, 306, 425),
+        d29x90(271, 29, 90, false, 306, 991),
+        d38x90(272, 38, 90, false, 413, 991),
+        d39x48(367, 39, 48, false, 425, 495),
+        d52x29(374, 52, 29, false, 578, 271),
+        d60x86(383, 60, 86, false, 672, 954),
+        d62x29(274, 62, 29, false, 696, 271),
+        d62x100(275, 62, 100, false, 696, 1109),
+        d102x51(365, 102, 51, false, 1164, 526),
+        d102x152(366, 102, 152, false, 1164, 1660),
+        d103x164(385, 103, 164, false, 1200, 1822);
+
+        override fun toString(): String {
+            var n = if (this.continuous) "${this.width} mm" else "${this.width} mm × ${this.height} mm"
+            if (this.twoColor) {
+                n += " (red/black)"
+            }
+            return n
+        }
+    }
+
+    override fun allowedForUsecase(type: String): Boolean {
+        return type != "receipt"
+    }
+
+    override fun convertPageToBytes(img: Bitmap, isLastPage: Boolean, previousPage: Bitmap?, conf: Map<String, String>, type: String): ByteArray {
+        val ostream = ByteArrayOutputStream()
+
+        val label = Label.values().find { it.name == conf.get("hardware_${type}printer_label") }!!
+        val targetWidth = label.printableWidth
+
+        val matrix = Matrix()
+        var imgW = img.width
+        var imgH = img.height
+        // rotate
+        if (conf.get("hardware_${type}printer_rotate90") == "true") {
+            matrix.setRotate(90f)
+            imgH = img.width
+            imgW = img.height
+        }
+        // resize
+        var targetHeight = imgH
+        if (imgW > targetWidth) {
+            targetHeight = (targetWidth.toFloat() / imgW.toFloat() * imgH.toFloat()).toInt()
+            val sx: Float = targetWidth / imgW.toFloat()
+            val sy: Float = targetHeight / imgH.toFloat()
+            matrix.postScale(sx, sy)
+        }
```

Comment:
Some of our backends (FGL, SLCS) do not have a resize step but instead just output the picture 1:1 in the size that is specified in the pretix backend. This has the advantage, that 1cm in the pretix ticket editor is really 1cm on paper later, but the downside that I get a broken print if I set the wrong sizes. Your approach (which we apparently also did in GraphicESCPOS) "just works", which is nice.

I don't argue that either one is better and I think I'm fine if we continue handling this inconsistently. The reason for the inconsistency is mostly because with GraphicESCPOS and with Brother, we need to know what size of paper we have anyways, while with FGL or SLCS we have no idea and currently do not offer special configuration for it, so it's like we can easily change that.

However, there's one thing to consider here (which I apparently did not consider with GraphicESCPOS): DPI and double rendering. Given a DPI of 300, `renderFileTo` will already render the pdf file to a 300dpi bitmap using the paper size specified in pretix' backend. Now, assuming that the pretix backend file is maybe a few millimeters too large, your code is taking that bitmap file, and resizing it *again* to the *almost same size*. I am a little worried that this "double resize" with a scaling factor very close to 1 leads to situations where the cubic filter used for resizing by createBitmap might loose pixels in a small QR code due to rounding effects which could harm readability of the barcode.

Do you think that my hunch is correct here or am I overreacting?

A simple solution could be to always set the DPI for the internal rendering to e.g. 600dpi and then scale down from there, a scale factor of ~2 should work better than "almost 1".

Reply:
Hmm. I've just tested it with a little too big ticket template (31mm template → 29mm paper), so the double resize happened. Internal rendering with 300dpi and 600dpi shows no visible difference, also very small qr codes (~4mm) are scanned without problems by the TC26 imager.

For me, simply increasing to 600dpi to be on the safe side, would be okay :)